Terraform crash during plan

### Terraform Version

```shell
Terraform v1.15.0
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v5.100.0
+ provider registry.terraform.io/hashicorp/external v2.3.5
+ provider registry.terraform.io/oracle/oci v5.47.0
```

### Terraform Configuration Files

```terraform
internal repository that we cannot share.
```


### Debug Output

```
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.
When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.
[1]: https://github.com/hashicorp/terraform/issues
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
panic: must not call MapVal with empty map
goroutine 2359 [running]:
runtime/debug.Stack()
	runtime/debug/stack.go:26 +0x5e
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	github.com/hashicorp/terraform/internal/logging/panic.go:84 +0x16a
panic({0x3be11a0?, 0x4d85770?})
	runtime/panic.go:783 +0x132
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1.1()
	github.com/hashicorp/terraform/internal/terraform/graph.go:59 +0x4c5
panic({0x3be11a0?, 0x4d85770?})
	runtime/panic.go:783 +0x132
github.com/zclconf/go-cty/cty.MapVal(0x6?)
	github.com/zclconf/go-cty@v1.18.0/cty/value_init.go:211 +0x485
github.com/hashicorp/terraform/internal/terraform.(*evaluationStateData).GetModule(0xc003299d40, {{}, {0xc000f014b8?, 0x0?}}, {{0xc000bc1f00, 0x33}, {0x2, 0xb, 0x28}, {0x2, ...}})
	github.com/hashicorp/terraform/internal/terraform/evaluate.go:596 +0x1279
github.com/hashicorp/terraform/internal/lang.(*Scope).evalContext(0xc00444fc20, {0xc0025dfa58, 0x1, 0x1}, {0x0, 0x0})
	github.com/hashicorp/terraform/internal/lang/eval.go:392 +0x1c63
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalContext(...)
	github.com/hashicorp/terraform/internal/lang/eval.go:248
github.com/hashicorp/terraform/internal/lang.(*Scope).EvalExpr(0xc00444fc20, {0x4df06f8, 0xc000dc0780}, {{0x4df0cd8?, 0x73c3340?}})
	github.com/hashicorp/terraform/internal/lang/eval.go:173 +0x8c
github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateExpr(0xc002767248?, {0x4df06f8, 0xc000dc0780}, {{0x4df0cd8?, 0x73c3340?}}, {0x0?, 0x0?})
	github.com/hashicorp/terraform/internal/terraform/eval_context_builtin.go:353 +0xab
github.com/hashicorp/terraform/internal/terraform.evalOutputValue({0x4e2c420, 0xc003f6c5a0}, {0x4df06f8, 0xc000dc0780}, {{0x4df0cd8?, 0x73c3340?}}, 0x0)
	github.com/hashicorp/terraform/internal/terraform/node_output.go:557 +0x8c
github.com/hashicorp/terraform/internal/terraform.(*NodeApplyableOutput).Execute(0xc0032a9200, {0x4e2c420, 0xc003f6c5a0}, 0x58?)
	github.com/hashicorp/terraform/internal/terraform/node_output.go:453 +0x6e6
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc0065454a0, {0x4e2c420, 0xc003f6c5a0}, {0x7f4ca3a040c8, 0xc0032a9200})
	github.com/hashicorp/terraform/internal/terraform/graph_walk_context.go:170 +0xaf
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x430bba0, 0xc0032a9200})
	github.com/hashicorp/terraform/internal/terraform/graph.go:143 +0x755
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc0068a9560, {0x430bba0, 0xc0032a9200}, 0xc002dfce00)
	github.com/hashicorp/terraform/internal/dag/walk.go:393 +0x2d1
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 1671
	github.com/hashicorp/terraform/internal/dag/walk.go:316 +0xf13
Error: Terraform exited with code 11.
╷
│ Error: Failed to read the given file as a state or plan file
│ 
│ State read error: Error loading statefile: open tfplan: no such file or
│ directory
│ 
│ Plan read error: couldn't load the provided path as either a local plan
│ file (open tfplan: no such file or directory) or a saved cloud plan (open
│ tfplan: no such file or directory)
╵
```


### Expected Behavior

Plan should have reached the phase where it shows the changes. 

### Actual Behavior

Terraform crashed during state refresh.
````
You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.
If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
Success! The configuration is valid.

module.cloud_config.data.oci_objectstorage_object.aws_cloud_network[0]: Reading...
module.cloud_config.data.oci_objectstorage_object.azure_cloud_network[0]: Reading...
module.cloud_config.data.oci_objectstorage_object.azure_subscriptions[0]: Reading...
module.cloud_config.data.oci_objectstorage_object.cloud_network[0]: Reading...
module.cloud_config.data.oci_objectstorage_object.aws_accounts[0]: Reading...
module.cloud_config.data.oci_objectstorage_object.oci_compartments[0]: Reading...
module.cloud_config.data.oci_objectstorage_object.aws_cloud_network[0]: Read complete after 0s
module.cloud_config.data.oci_objectstorage_object.azure_cloud_network[0]: Read complete after 0s
module.cloud_config.data.oci_objectstorage_object.cloud_network[0]: Read complete after 0s
module.cloud_config.data.oci_objectstorage_object.aws_accounts[0]: Read complete after 0s [id=ObjectStorageObjectDataSource-2116639164]
module.cloud_config.data.oci_objectstorage_object.oci_compartments[0]: Read complete after 0s [id=ObjectStorageObjectDataSource-1051408083]
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
````


### Steps to Reproduce

1. terraform init
2. terraform plan

### Additional Context

Terraform is executed by a GitHub workflow on a GitHub runner. Other jobs using the same workflow are executed without errors.

### References

_No response_

### Generative AI / LLM assisted development?

_No response_

--- a/.changes/v1.15/BUG FIXES-20260430-103820.yaml
+++ b/.changes/v1.15/BUG FIXES-20260430-103820.yaml
@@ -1,0 +1,5 @@
+kind: BUG FIXES
+body: Fix panic for types modules with no expanded instances
+time: 2026-04-30T10:38:20.648469-04:00
+custom:
+    Issue: "38491"

--- a/internal/terraform/evaluate.go
+++ b/internal/terraform/evaluate.go
@@ -437,23 +437,25 @@ func (d *evaluationStateData) GetModule(addr addrs.ModuleCall, rng tfdiags.Sourc
 		noDynamicTypes = noDynamicTypes && !out.ConstraintType.HasDynamicTypes()
 	}
 
-	if d.Operation == walkValidate && typeDefined {
-		atys := make(map[string]cty.Type, len(outputConfigs))
-		as := make(map[string]cty.Value, len(outputConfigs))
-		for name, c := range outputConfigs {
-			// atys is used to create the module object type for expanded modules
-			atys[name] = c.ConstraintType
-			// the unknown val can be used when we return a single module
-			// instance with unknown outputs
-			val := cty.UnknownVal(c.ConstraintType)
+	// build up the type of the configured module output object
+	atys := make(map[string]cty.Type, len(outputConfigs))
+	// and create a single unknown instance value for validation
+	as := make(map[string]cty.Value, len(outputConfigs))
+	for name, c := range outputConfigs {
+		// atys is used to create the module object type for expanded modules
+		atys[name] = c.ConstraintType
 
-			if c.DeprecatedSet {
-				val = val.Mark(marks.NewDeprecation(c.Deprecated, absAddr.Output(name).ConfigOutputValue().ForDisplay()))
-			}
-			as[name] = val
+		// the unknown val can be used when we return a single module
+		// instance with unknown outputs
+		val := cty.UnknownVal(c.ConstraintType)
+		if c.DeprecatedSet {
+			val = val.Mark(marks.NewDeprecation(c.Deprecated, absAddr.Output(name).ConfigOutputValue().ForDisplay()))
 		}
-		instTy := cty.Object(atys)
+		as[name] = val
+	}
+	instTy := cty.Object(atys)
 
+	if d.Operation == walkValidate && typeDefined {
 		switch {
 		case callConfig.Count != nil && noDynamicTypes:
 			return cty.UnknownVal(cty.List(instTy)), diags
@@ -579,9 +581,12 @@ func (d *evaluationStateData) GetModule(addr addrs.ModuleCall, rng tfdiags.Sourc
 			elems = append(elems, instVal)
 			diags = diags.Append(moreDiags)
 		}
-		if noDynamicTypes {
+		switch {
+		case noDynamicTypes && len(elems) == 0:
+			return cty.ListValEmpty(instTy), diags
+		case noDynamicTypes:
 			return cty.ListVal(elems), diags
-		} else {
+		default:
 			return cty.TupleVal(elems), diags
 		}
 
@@ -592,9 +597,13 @@ func (d *evaluationStateData) GetModule(addr addrs.ModuleCall, rng tfdiags.Sourc
 			attrs[string(instKey.(addrs.StringKey))] = instVal
 			diags = diags.Append(moreDiags)
 		}
-		if noDynamicTypes {
+
+		switch {
+		case noDynamicTypes && len(attrs) == 0:
+			return cty.MapValEmpty(instTy), diags
+		case noDynamicTypes:
 			return cty.MapVal(attrs), diags
-		} else {
+		default:
 			return cty.ObjectVal(attrs), diags
 		}
 

--- a/internal/terraform/evaluate_test.go
+++ b/internal/terraform/evaluate_test.go
@@ -612,6 +612,14 @@ func TestEvaluatorGetModule_validateTypedOutputs(t *testing.T) {
 				"out": cty.String,
 			}))),
 		},
+		"empty_count": {
+			configureModuleCall: func(call *configs.ModuleCall) {
+				call.Count = hcltest.MockExprLiteral(cty.NumberIntVal(0))
+			},
+			want: cty.UnknownVal(cty.List(cty.Object(map[string]cty.Type{
+				"out": cty.String,
+			}))),
+		},
 		"for_each": {
 			configureModuleCall: func(call *configs.ModuleCall) {
 				call.ForEach = hcltest.MockExprLiteral(cty.MapVal(map[string]cty.Value{
@@ -729,6 +737,15 @@ func TestEvaluatorGetModule_planTypedOutputs(t *testing.T) {
 				cty.ObjectVal(map[string]cty.Value{"out": cty.StringVal("second").Mark(marks.Sensitive)}),
 			}),
 		},
+		"empty_count": {
+			setupInstances: func(expander *instances.Expander) {
+				expander.SetModuleCount(addrs.RootModuleInstance, addrs.ModuleCall{Name: "mod"}, 0)
+			},
+			setupOutputs: func(namedValues *namedvals.State) {
+				// explicitly setting no values
+			},
+			want: cty.ListValEmpty(cty.Object(map[string]cty.Type{"out": cty.String})),
+		},
 		"for_each": {
 			setupInstances: func(expander *instances.Expander) {
 				expander.SetModuleForEach(addrs.RootModuleInstance, addrs.ModuleCall{Name: "mod"}, map[string]cty.Value{
@@ -754,6 +771,15 @@ func TestEvaluatorGetModule_planTypedOutputs(t *testing.T) {
 				"a": cty.ObjectVal(map[string]cty.Value{"out": cty.StringVal("first").Mark(marks.Sensitive)}),
 				"b": cty.ObjectVal(map[string]cty.Value{"out": cty.StringVal("second").Mark(marks.Sensitive)}),
 			}),
+		},
+		"empty_for_each": {
+			setupInstances: func(expander *instances.Expander) {
+				expander.SetModuleForEach(addrs.RootModuleInstance, addrs.ModuleCall{Name: "mod"}, map[string]cty.Value{})
+			},
+			setupOutputs: func(namedValues *namedvals.State) {
+				// no values for empty for_each
+			},
+			want: cty.MapValEmpty(cty.Object(map[string]cty.Type{"out": cty.String})),
 		},
 	}
 

